### ISSUE TITLE
Faster processing of `Data` objects

### ISSUE THREAD

**Issue by effectfully** [IntersectMBO/plutus#6225]
## Preface

This issue is a continuation of #5777 (I'm going to call it "the previous issue") and offers a different perspective. Reading the previous issue is a prerequisite for reading this one.

The previous issue explained how it'd be beneficial to have a way of matching on `Data` objects by converting them to SOPs first: this would make processing of `Data` much more efficient. The issue previous also explaines how there doesn't appear to exist any safe and efficient way to perform the conversion. Hence here we're going to investigate alternatives.

## The goal

We want to be able to express the following (pseudocode):

```
inspectConstr (Constr 2 [d0, d1, d2])
    [ \[v0, v1]         -> body0
    , \[v0, v1, v2, v3] -> body1
    , \[v0, v1, v2]     -> body2
    , \[]               -> body3
    ]
```

where `inspectConstr` is some function that takes a `Data` object and a bunch of branches (in the form of a list or an array or something), picks the right branch depending on the index of the constructor (i.e. the first argument of `Constr`) and applies the multi-argument function of that branch to the arguments of the constructor (i.e. what's stored in the second argument of `Constr`). So the example above evaluates to

```
(\[v0, v1, v2] -> body2) [d0, d1, d2]
```

where `\[v0, v1, v2]` is a multi-lambda binding three variables and `[d0, d1, d2]` is some kind of a spine (a list or an array or etc, can be the same thing as what we store the branches in, can be different). That expression is equivalent to

```
(\v0 v1 v2 -> body2) d0 d1 d2
```

and evaluates the same way.

Note that regardless of how we're going to represent this expression in the AST we want to only allow exactly-saturated applications for multi-lambdas, i.e. neither undersaturation (a multi-lambda binds more variables than the number of arguments it's applied to) nor oversaturation (a multi-lambda binds fewer variables than the number of arguments it's applied to) are allowed. The reason for this restriction is precisely what allows us to avoid all the problems described in the "What exactly breaks if we add the unsafe `constrTermFromConstrData`?" section of the previous issue.

<to_be_continued>

**Comment by MicroProofs** [comment]
So switching gears to multi-apply and multi-lambda and multi-apply and also adding arrays as a plc term right? Just to summarize this PR.

**Comment by effectfully** [comment]
> So switching gears to multi-apply and multi-lambda and multi-apply and also adding arrays as a plc term right? Just to summarize this PR.

Yes, so far this seems like the most plausible solution to me (just to clarify, it's an issue, not a PR, I haven't written any code for that quite yet), but others may have better ideas and I'm not entirely confident the idea is gonna work out and be worth it.

**Comment by effectfully** [comment]
@IntersectMBO/plutus-core please review this issue carefully and check whether there are any flaws in my reasoning or whether there's something missing. If I don't get any negative feedback, I'm going to proceed with shipping pattern matching builtins instead of the SOPs approach (see [this](https://github.com/IntersectMBO/plutus/issues/5711) issue for details on what the two approaches are).

**Comment by kwxm** [comment]
> If I don't get any negative feedback, I'm going to proceed with shipping pattern matching builtins

OK, go ahead.  I don't think I fully appreciate all of the nuances of the different approaches, but we've talked about his quite a lot and I think you've convinced me that your plan is sensible.  It may complicate the specification and the metatheory, but I suppose we'll just have to deal with that.

I suppose we'll have to show that this gives decent gains on realistic code, so we should probably try it out on some big examples.  Maybe once you have a working implementation we can get some community members to try it out for us and see how much help it is.  Do we know how complicated the objects are that people are working with?  If you're trying to get at something buried deeply inside a `Data` object then you'd need multiple applications of this stuff.  Thinking randomyl,  would it make any sense to think about some atomic operation that takes path through a `Data` object and extracts the thing that it leads to?  Maybe that wouldn't be helpful if you need to get at multiple things.

**Comment by effectfully** [comment]
(this issue only tangentially concerns pattern matching builtins, [that](https://github.com/IntersectMBO/plutus/issues/5711) one is specifically about those, but I'll respond here)

>  It may complicate the specification and the metatheory,

I'm sure it'll complicate the former, but for the latter I feel like the difference isn't very substantial?

> Maybe once you have a working implementation we can get some community members to try it out for us and see how much help it is.

That makes sense, but it'll further delay shipment. @zliu41 what do you think? We do [know](https://github.com/IntersectMBO/plutus/pull/5486) that pattern matching builtins speed up list-processing microbenchmarks by 25-30% and data-processing microbenchmarks by 15-20%, both in the presence of recursion, i.e. these are quite realistic microbenchmarks. And our "pickiness standards" were lower for, say, SOPs, for which we were fine with less pronounced microbenchmarking [results](https://github.com/michaelpj/CIPs/blob/192e14988bdccb600450e6db09977d9ed801ebd2/CIP-0085/README.md#appendix-3-benchmark-results) and a 4% slowdown of existing code not using the feature, while pattern matching builtins don't introduce any measurable slowdown. 

(BTW, if anybody's interested why lists are optimized more, my conjecture is that it's the cost of lists being a polymorphic built-in type, so calling list-handling builtins fewer times gives a larger speedup than calling data-handling builtins fewer times. And the reason why polymorphic built-in types are slower is because I designed the machinery the wrong way and can't get to fixing it).

> Thinking randomyl, would it make any sense to think about some atomic operation that takes path through a Data object and extracts the thing that it leads to? Maybe that wouldn't be helpful if you need to get at multiple things.

... in which case we may just choose to return a list or even a subdata. The user will then still need to process the result, but at least they won't pay much for skipping the redundant bits with that "subdata" approach. Sounds interesting, we should perhaps look into it. It also sounds a bit like a specialized version of [folding builtins](https://github.com/IntersectMBO/plutus/issues/6034) (assuming we can make those lazy enough somehow).

**Comment by zliu41** [comment]
Have you considered updating all `UnsafeFromData` instances to use `matchData`? That should lead to cheaper decoding of things, especially `ScriptContext`, and it would be interesting to see how much difference it makes.

**Comment by effectfully** [comment]
> Have you considered updating all `UnsafeFromData` instances to use `matchData`? That should lead to cheaper decoding of things, especially `ScriptContext`, and it would be interesting to see how much difference it makes.

(I've done that in the pattern matching builtins [PR](https://github.com/IntersectMBO/plutus/pull/6530))

**Comment by effectfully** [comment]
I've been thinking about this and thing I didn't consider before is that in

```haskell
inspectConstr (Constr 2 [d0, d1, d2])
    [ \[v0, v1]         -> body0
    , \[v0, v1, v2, v3] -> body1
    , \[v0, v1, v2]     -> body2
    , \[]               -> body3
    ]
```

the array of functions must store them lazily, otherwise lookup is linear, which is pointless. But making arrays lazy is really weird, particularly given that we don't have call-by-need, just call-by-name.

On the other hand, `case`-expressions are already appropriately lazy...
